### DISCUSSION TITLE
ArceOS USB 3.0 驱动开发

### DISCUSSION THREAD

**Discussion by Rionzh** [chenlongos/usb-doc#1] [Announcements]
# ArceOS USB 3.0 驱动开发

## 基本要求

1. 要达到以下要求才可正式加入到 USB 3.0 驱动开发的小组中（即进入微信交流小群）：
   * 了解掌握现有进展下的代码（<https://github.com/dbydd/arceos_experiment/tree/task3_usb>）
   * 主函数test_usb_driver（），位置在apps/cli/src/cmd.rs
   * crates：brcm_pcie、driver_xhci
2. 小组要求：每周由同学们进行一个简单的技术分享
   * 具体分享时间（群里通知）

## 开发指南

以树莓派4B开发板为例，在其进行 USB 3.0 驱动开发，首先：

1. 完成 PCIe 主桥的初始化工作，使得 PCIe 主桥可以正常工作
   * 每个 PCIe 主桥都有一组配置寄存器，用于配置和管理主桥的各种功能。这些寄存器通常由 MMIO（内存映射 I/O）空间提供访问。确定这些寄存器的基地址和偏移量，进行初始化和配置。

2. PCIe 主桥枚举连接到总线上的设备，发现 USB 3.0 主机控制器设备并解析其配置空间(PCI Express Configuration Space)
   * 执行 PCIe 总线扫描以检测连接到总线上的所有设备。通常通过读取配置空间中的特定寄存器来实现。
   * 对于检测到的设备，需要解析其配置空间以获取设备的厂商 ID、设备 ID、类别码等重要信息。
   * USB 3.0 主机控制器设备的类别码通常是 0x0C（Serial bus controller）。

3. 对发现的USB3.0主机控制器分配内存空间
   * 使用适当的内存管理函数进行内存分配。
   * 分配物理内存：对于 PCIe 设备，通常需要分配物理内存，而不是虚拟内存。
   * 将物理内存映射到虚拟地址空间。

4. USB 3.0 主机控制器的初始化，使得主机控制器可以正常工作
   * 设置主机控制器寄存器。
   * 配置主机控制器操作模式：选择适当的主机控制器操作模式。包括设置传输速度（例如超速或高速）、电源管理模式、端点配置等。
   * 初始化传输环路和命令队列：USB 3.0 主机控制器使用传输环路和命令队列来管理数据传输和控制请求。需要初始化这些数据结构，并根据需要配置它们的大小、数量和优先级。
   * 配置中断处理：USB 3.0 主机控制器会产生各种类型的中断，例如传输完成、错误报告、连接状态变化等。
   * 分配和初始化数据缓冲区：USB 3.0 主机控制器使用数据缓冲区来存储传输的数据。需要分配适当大小的数据缓冲区，并进行初始化以确保正确的数据传输和接收。
   * 配置端点参数：USB 3.0 主机控制器使用端点来管理数据传输的方向、类型和属性。需要配置每个端点的传输参数，例如最大包大小、传输方向、间隔等。
   * 启用主机控制器功能：在完成上述配置后，设置相应的寄存器位或调用特定的函数来启用主机控制器。
5. USB主机控制器枚举和检测插入的 USB 设备
6. 创建适当的数据结构和数据通路来管理和处理 USB 设备的连接和通信
   * 设备描述符解析：当 USB 3.0 设备连接到主机控制器时，需要解析设备描述符以获取设备的基本信息，如厂商 ID、产品 ID、设备类别等。
   * 端点管理：USB 3.0 设备使用端点来进行数据传输和通信。创建适当的数据结构来管理每个端点，包括输入输出端点的类型、方向、最大包大小等信息。
   * 传输请求处理：在与 USB 3.0 设备通信时，需要处理传输请求，包括执行控制传输、批量传输、中断传输和等时传输等。
   * 错误处理和调试：在开发 USB 3.0 主机控制器驱动程序时，错误处理和调试是至关重要的。
7. 实现具体的USB设备驱动程序
   * 确定设备类型和功能：确定具体 的USB 设备类型和功能。包括存储设备、音频设备、网络设备等。
   * 创建驱动程序框架：创建驱动程序框架。包括初始化和配置驱动程序，注册设备与驱动程序之间的关联，设置回调函数等。
   * 实现设备插拔处理：涉及监测设备的插入和拔出事件，识别设备并分配必要的资源，建立连接和通信。
   * 数据传输与控制：实现设备与主机之间的数据传输和控制功能，包括读取和写入数据、发送和接收命令以及管理设备的状态等操作。使用适当的 USB 协议和传输方式，确保与设备的正确通信。
   * 错误处理与调试：实现错误处理和调试功能，以便在运行时可以检测和处理各种错误情况，包括错误码处理、日志记录、异常条件处理等。

## 进展分享（树莓派）

* 许慎林完成了第一步：PCIe主桥的初始化工作，使得PCIe主桥可以正常工作。（<https://github.com/Axsl666/arceos>）

* 姚宏伟完成了第二、三、四步，包括：PCIe主桥枚举连接到总线上的设备，发现USB3.0主机控制器设备、对发现的USB3.0主机控制器分配内存空间、USB3.0主机控制器的初始化，使得主机控制器可以正常工作。（<https://github.com/dbydd/arceos_experiment/tree/task3_usb>）

## 资料分享（[树莓派项目资料 ](https://docs.qq.com/doc/DSU1QalVwWmtuT3NF)）

1. [实验指导书](https://chenlongos.com/raspi4-with-arceos-doc/)
2. [ArceOS环境配置](https://rcore-os.cn/arceos-tutorial-book/ch01-00.html)
3. [树莓派新手入门手册](https://github.com/chenlongos/raspi4-with-arceos-doc/blob/master/src/assert/树莓派4B新手入门手册.pdf)
4. [bcm2711芯片手册](https://datasheets.raspberrypi.com/bcm2711/bcm2711-peripherals.pdf)
5. [VL805相关数据手册](https://github.com/chenlongos/raspi4-with-arceos-doc/blob/master/src/assert/DS_VLI_VL805_093.pdf)
6. [xhci协议手册](https://github.com/chenlongos/raspi4-with-arceos-doc/blob/master/src/assert/extensible-host-controler-interface-usb-xhci.pdf)
7. [**redox**（作为参考的x86_OS）](https://github.com/redox-os/drivers)
8. 2023.11.30分享交流会视频回放：[开源操作系统培训 - 项目五：ChenLongOS辰龙操作系统发行版 - os2edu.cn](https://os2edu.cn/course/161)
9. 2023.11.30分享交流会PPT：（标题链接里下载）

（1）关于在树莓派上进行arceos usb驱动开发的进展-1.pptx（许慎林）

（2）关于在树莓派上进行arceos usb驱动开发的进展-2.pptx（姚宏伟）

持续更新中🚗🚗......

**Comment by XxChang** [comment]
分享一个USB开发的资料网站[USB in a NutShell](https://beyondlogic.org/usbnutshell/usb1.shtml)
